### ISSUE TITLE
DCPMM becomes a jbd file system after reboot

### ISSUE THREAD

**Issue by davidliu2012** [intel/ipmctl#106]
I have installed two 128GB DCPMMs (AppDirect, 1-1-1) on my server. I use Ubuntu 18.04 (kernel 4.18) as my OS. My impctl version number is "01.00.00.3279", and memory FWVersion is "01.02.00.5355".
Initially, I created a fsdax namespace with ndctl (61.2+) and created an ext4 fs on top with "mkfs.ext4 /dev/pmem0". It was fine at first. However, one day I found the device pmem0 could not be mounted when I ran "mount -o dax /dev/pmem0 /mnt/pmem0". The error message was "**mount: /mnt/pmem0: unknown filesystem type 'jbd'.**" I am wondering if that is a data corruption (ipmctl shows my DCPMMs are Healthy) or I used the wrong command. Thanks!

**Comment by sscargal** [comment]
Yes, that sounds like something overwrote critical ext4 metadata and it just so happens that whatever is there now maps to a 'jbd' value which is what is being returned.  What does `tune2fs -l /dev/pmem0` return?  If it's garbage, then you know the superblock and other data has been mangled.  You can use `dd`, `od, or `xxd` to dump the first few megabytes of the `/dev/pmem0` device to see if you/we can recognise anything that may have written over the region.

Q) What have you been doing the `/dev/pmem0` or `/mnt/pmem0`?  (specifically what has been writing)

**Comment by davidliu2012** [comment]
Thanks for replying! I reinitialized the file system after that. I have had the same problem a couple of times already. I think it might have something to do with the OS. I will try these commands once I encountered the same problem again.

**Comment by davidliu2012** [comment]
I found the reason why the drive corrupted. This happens only when I reinstalled the CPU. Is this a protection feature? 

The following is the hexdump of the /dev/pmem0 header:
```

hexdump crash | head -n20
0000000 0000 0000 0000 0000 0000 0000 0000 0000
*
00000a0 aaaa aaaa aaaa aaaa 0000 0000 0000 0000
00000b0 0000 0000 0000 0000 0000 0000 0000 0000
*
0000100 aaaa aaaa aaaa aaaa 0000 0000 0000 0000
0000110 0000 0000 0000 0000 0000 0000 0000 0000
*
00001c0 aaaa aaaa aaaa aaaa 0000 0000 0000 0000
00001d0 0000 0000 0000 0000 0000 0000 0000 0000
*
0000220 aaaa aaaa aaaa aaaa 0000 0000 0000 0000
0000230 0000 0000 0000 0000 0000 0000 0000 0000
*
00002e0 aaaa aaaa aaaa aaaa 0000 0000 0000 0000
00002f0 0000 0000 0000 0000 0000 0000 0000 0000
*
0000340 aaaa aaaa aaaa aaaa 0000 0000 0000 0000
0000350 0000 0000 0000 0000 0000 0000 0000 0000
*
```

**Comment by sscargal** [comment]
Thanks!

I agree this is not the expected on-disk layout contents for an ext4 filesystem.  I don't recognize `aaaa aaaa aaaa aaaa` as anything relating to persistent memory pools or ext4.  

It is supported to physically move Intel DC Optane Persistent Memory between systems, so we don't tie ourselves to the CPU.  Swapping or upgrading/downgrading CPUs between supported SKUs should work fine (I've done it several times).  

Q) What server are you using (OEM/ODM Manufacturer & model#)
Q) What exactly are you doing when you 'reinstall the CPU'?  Can you provide a step-by-step procedure for what you do to reproduce the issue please.  Are you reseating the same CPU or switching it out for another SKU?  
Q) Can you provide the model number (`lscpu`) for the CPUs you are using please
Q) Do the CPUs support DCPMM? (We can check using the model number)
Q) Have you enabled the passphrase (extra security) for DCPMM?
Q) Have you enabled LUKS or similar disk/filesystem encryption?
Q) Without changing HW, if you gracefully reboot or power-cycle the host, do you see the same issue?

**Comment by sihangliu** [comment]
1. I am using a Lenovo SR650 with a Xeon Gold 6230. It supports DCPMM.
2. The same problem occurred twice. 
The first time is I unplugged the CPU and changed its thermal paste. The file system on DCPMM corrupted after I boot the machine again. 
And the second time, I tested another CPU (Xeon Platinum 8260) on the server. The DCPMM data corrupted when I changed from 6230 to 8260 and from 8260 to 6230. 
Could that be a Lenovo feature that protects persistent data? I was told about this feature when ordering the server. 
3. I did not enable anything other than setting to AppDirect Mode. Is there a security option in ipmctl that can cause such an issue? I initialized the file systems with the commands in the first post. I didn't use any encryption option.
4. If I just shutdown the computer, unplug the power cable and power it back on, the file system on DCPMM is still mountable (no data loss). 

The following is `lscpu` from Xeon Gold 6230 (I don't have the 8260 in hand now).
```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              40
On-line CPU(s) list: 0-39
Thread(s) per core:  2
Core(s) per socket:  20
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz
Stepping:            7
CPU MHz:             800.139
BogoMIPS:            4200.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            28160K
NUMA node0 CPU(s):   0-39
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp_epp pku ospke avx512_vnni md_clear flush_l1d arch_capabilities

```

**Comment by sscargal** [comment]
> Could that be a Lenovo feature that protects persistent data? I was told about this feature when ordering the server.

I would create a support ticket with Lenovo to find out more.  If the DCPMM is linked to the CPUs, it means you would not be able to move CPUs between servers.  I haven't personally run into that issue when replacing CPUs on Intel white box servers and firmware.  I'll check to make sure.
